### PR TITLE
sniproxy: add promote-https

### PR DESCRIPTION
**Pull request by shayne**
Adds support for an HTTP server that promotes all requests to HTTPS.
The flag is `-promote-https` and defaults to true.

Updates #1748

### CODE REVIEW

File: cmd/sniproxy/snipproxy.go [hunk 1]
```diff
@@ -23,6 +24,7 @@ import (
 )
 
 var ports = flag.String("ports", "443", "comma-separated list of ports to proxy")
+var promoteHTTPS = flag.Bool("promote-https", false, "promote HTTP to HTTPS")
```
Comment (crawshaw) [inline cmd/sniproxy/snipproxy.go:27]: I'm not even sure it needs a flag, seems like always the right behavior. At least I'd default to true.

File: cmd/sniproxy/snipproxy.go [hunk 2]
```diff
@@ -197,3 +208,9 @@ func (s *server) dnsResponse(req *dnsmessage.Message) (buf []byte, err error) {
 
 	return resp.Finish()
 }
+
+func (s *server) promoteHTTPS(ln net.Listener) {
+	http.Serve(ln, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Comment (crawshaw) [inline cmd/sniproxy/snipproxy.go:213]: Good to log the error message that comes out of Serve

**Review by crawshaw — APPROVED** [review]

**Comment by DentonGentry** [comment]
Please add "Updates https://github.com/tailscale/tailscale/issues/1748" to the commit message.

File: cmd/sniproxy/snipproxy.go [hunk 3]
```diff
@@ -197,3 +208,10 @@ func (s *server) dnsResponse(req *dnsmessage.Message) (buf []byte, err error) {
 
 	return resp.Finish()
 }
+
+func (s *server) promoteHTTPS(ln net.Listener) {
+	err := http.Serve(ln, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		http.Redirect(w, r, "https://"+r.Host+r.RequestURI, http.StatusMovedPermanently)
+	}))
+	log.Printf("promoteHTTPS http.Serve: %v", err)
```
Comment (bradfitz) [inline cmd/sniproxy/snipproxy.go:216]: log.Fatalf?

File: cmd/sniproxy/snipproxy.go [hunk 4]
```diff
@@ -197,3 +208,10 @@ func (s *server) dnsResponse(req *dnsmessage.Message) (buf []byte, err error) {
 
 	return resp.Finish()
 }
+
+func (s *server) promoteHTTPS(ln net.Listener) {
+	err := http.Serve(ln, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		http.Redirect(w, r, "https://"+r.Host+r.RequestURI, http.StatusMovedPermanently)
```
Comment (bradfitz) [inline cmd/sniproxy/snipproxy.go:214]: I think StatusFound or something less permanent is safer cache-wise? I always have to look up the differences, so maybe I'm remembering wrong.

**Pull request merged**